tests/tls_e2e/client_enc/client.cpp

```diff
@@ -0,0 +1,433 @@
+    mbedtls_ssl_config* conf,
+    mbedtls_ctr_drbg_context* ctr_drbg,
+    mbedtls_x509_crt* client_cert,
+    mbedtls_pk_context* private_key)
+{
+    int ret = 1;
+    oe_result_t result = OE_FAILURE;
+
+    OE_TRACE_INFO("Generating the certificate and private key\n");
+    result = generate_certificate_and_pkey(client_cert, private_key);
+    if (result != OE_OK)
+    {
+        OE_TRACE_ERROR("failed with %s\n", oe_result_str(result));
+        ret = 1;
+        goto done;
+    }
+
+    OE_TRACE_INFO("Setting up the SSL/TLS structure...\n");
+
+    if ((ret = mbedtls_ssl_config_defaults(
+             conf,
+             MBEDTLS_SSL_IS_CLIENT,
+             MBEDTLS_SSL_TRANSPORT_STREAM,
+             MBEDTLS_SSL_PRESET_DEFAULT)) != 0)
+    {
+        OE_TRACE_ERROR(
+            "failed! mbedtls_ssl_config_defaults returned %d\n", ret);
+        goto done;
+    }
+
+    // set up random engine
+    mbedtls_ssl_conf_rng(conf, mbedtls_ctr_drbg_random, ctr_drbg);
+    // set debug function
+    mbedtls_ssl_conf_dbg(conf, debug_print, stdout);
+
+    // Customized certificate validation will replace TLS default operation, set
+    // authmod to OPTIONAL here Set the certificate verification mode to
+    // MBEDTLS_SSL_VERIFY_OPTIONAL, which peer certificate is checked, however
+    // the handshake continues even if verification failed;
+    // mbedtls_ssl_get_verify_result() can be called after the handshake is
+    // complete.
+    mbedtls_ssl_conf_authmode(conf, MBEDTLS_SSL_VERIFY_OPTIONAL);
+
+    mbedtls_ssl_conf_verify(conf, cert_verify_callback, NULL);
+
+    // enable TLS server to send a list of acceptable CAs in CertificateRequest
+    // messages. mbedtls_ssl_conf_cert_req_ca_list( &conf,
+    // MBEDTLS_SSL_CERT_REQ_CA_LIST_ENABLED); mbedtls_ssl_conf_ca_chain(conf,
+    // server_cert->next, NULL);
+
+    // Set own certificate chain and private key.
+    // Note
+    // own_cert should contain in order from the bottom up your certificate
+    // chain. The top certificate (self-signed) can be omitted. On server, this
+    // function can be called multiple times to provision more than
+    //            one cert/key pair (eg one ECDSA, one RSA with SHA-256, one RSA
+    //            with SHA-1). An adequate certificate will be selected
+    //            according to the client's advertised capabilities. In case
+    //            multiple certificates are adequate, preference is given to the
+    //            one set by the first call to this function, then second, etc.
+    // On client, only the first call has any effect. That is, only one client
+    // certificate
+    //            can be provisioned. The server's preferences in its
+    //            CertficateRequest message will be ignored and our only cert
+    //            will be sent regardless of whether it matches those
+    //            preferences - the server can then decide what it wants to do
+    //            with it.
+    //
```

Comment:
Formatting seems OFF - Is this intentional?

Reply:
Fixed